### PR TITLE
[Hackney] Alloy environmental services integration

### PR DESCRIPTION
**Pull request by davea**
Some changes/improvements to support the Alloy environmental services integration launch.

[skip changelog]

### CODE REVIEW

File: perllib/FixMyStreet/Cobrand/Hackney.pm [hunk 1]
```diff
@@ -243,6 +243,13 @@ sub munge_sendreport_params {
     }
 }
 
+sub open311_munge_update_params {
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Hackney.pm:299]: I was going to say there is a FixMyStreet::Roles::Open311Multi you can include, but I guess that doesn't work unless Hackney had a function already, hmm. Perhaps the role function could be a normal function that calls super instead? Doesn't seem to need to be an "around" way of doing it, I mean.

File: perllib/FixMyStreet/Cobrand/Hackney.pm [hunk 2]
```diff
@@ -143,6 +143,14 @@ around open311_extra_data_include => sub {
         $contact->email($code) if $code;
     }
 
+    if ($row->geocode) {
+        my $address = $row->geocode->{resourceSets}->[0]->{resources}->[0]->{address};
+        push @$open311_only, (
+            { name => 'closest_address', value => $address->{formattedAddress} }
+        );
+        $h->{closest_address} = '';
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Hackney.pm:151]: Ah, clever. (CheshireEast/Oxfordshire/Rutland do not do this, wonder if they should? Or if it's working anyway somehow.)

File: perllib/FixMyStreet/Cobrand/Hackney.pm [hunk 3]
```diff
@@ -251,6 +251,41 @@ sub munge_sendreport_params {
...
+    # have manually-created extra fields that should be appended to the
+    # report title when submitted via Open311.
+    # This fetches the field names from the COBRAND_FEATURES config, so
+    # if they change in the future it doesn't require code changes to deploy.
+    my $field_names = $self->feature('environment_extra_fields') || [];
+    my %fields = map { $_ => 1 } @$field_names;
+    if ($contact->email =~ /^Environment/) {
+        for (@{ $row->get_extra_fields }) {
+            if ($fields{$_->{name}}) {
+                $title .= "\n\n" . $_->{description} . "\n" . $_->{value};
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Hackney.pm:269]: Could do with a test of this.

File: perllib/FixMyStreet/Cobrand/Hackney.pm [hunk 4]
```diff
@@ -282,6 +282,16 @@ sub open311_extra_data_include {
           value => $row->category },
     ];
 
+    if ($contact->email eq "Environment-Graffiti") {
+        push @$open311_only,
+            { name => 'graffiti_offensive', value => $row->get_extra_field_value('offensive') || '' };
```
Comment (dracos) [inline perllib/FixMyStreet/Cobrand/Hackney.pm:287]: Could the extra field values match the names needed by open311 adapter to save this being needed? (Or the names in open311-adapter changed to match the FMS values?) Or am I missing something?

**Review by dracos — CHANGES_REQUESTED** [review]
Only a few small requests/queries again :)

File: perllib/FixMyStreet/Cobrand/Hackney.pm (hunk 4 above)
Comment (davea) [reply perllib/FixMyStreet/Cobrand/Hackney.pm]: Yeah I'm not sure what my thinking was, in hindsight. Think I wanted to preserve the field names from live for some reason but make it obvious what they were for in open311-adapter. Anyway have reverted this and used the `graffiti_`-prefixed field names in FMS.

**Review by dracos — APPROVED** [review]

**Pull request merged**